### ISSUE TITLE
Weird shading in FOG, ALPHA_TEST region

### ISSUE THREAD

**Issue by devendrn** [devendrn/newb-shader-mcbe#7]
Some devices seems to be having issues with ALPHA_TEST (leaves, grass, plants, torch, ladder, et..) objects and grass block color in FOG region. For ALPHA_TEST objects it only happens when PLANTS_WAVE is enabled, and when its raining.

This bug was not present in 1.1.8 
This bug doesn't happen in 1.1.9 lite

![Screenshot_20211124_182904_com mojang minecraftpe](https://user-images.githubusercontent.com/91605478/143288411-4e62157f-8aeb-4f38-be98-e05017dc608b.jpg)

**Comment by devendrn** [comment]
Update:
The issue has been narrowed down to these codes of terrain.fragment thanks to JulioCT
```
#ifdef FOG
	// wet effect
	if(wetRefl.a > 0.0){
		// only on xz plane
		if(isTopSurface(wPosY)){
			float mask = wetRefl.a*max((clamp(shade,0.78,0.88)-0.78)/0.1,0.4);
			diffuse.rgb = diffuse.rgb*(1.0 - 0.5*mask) + wetRefl.rgb*mask;
		}
	}
#endif
```
```
// color correction
diffuse.rgb = colorCorrection(diffuse.rgb);
```
The issue seems to be related to precision. This can be fixed soon. The wet reflection code is somehow messing the precision of `diffuse` which on color correction becomes weird.

**Comment by devendrn** [comment]
Apparently it was caused because of defining a bool function for finding top surface. It has been fixed in commit a62a041

**Comment by FasterMars** [comment]
This is one of the most detailed Github issues I've ever seen
